Merge entrypoints

danakim opened the pull request.

quartje reviewed (commented):
I think that the original CMD and ENTRYPOINT seperation might be good to keep (and make this change as small as possible)

danakim commented on php72-apache2/bin/start.sh:
```diff
@@ -0,0 +1,64 @@
+fi
+
+# Start Apache2
+apache2 -D FOREGROUND
```
Same like my comment above :)

quartje commented on php72-apache2/Dockerfile:
```diff
@@ -25,28 +25,30 @@ RUN docker-php-ext-install soap \
-COPY ./conf/entrypoint.sh /entrypoint.sh
-RUN chmod 755 /entrypoint.sh
+# Copy the startup script
+COPY ./bin/start.sh /usr/local/bin/start.sh
```
No chmod here?

danakim replied:
No, not needed as long as the script is already executable in the repo. And it is :)

quartje reviewed (changes requested):
I think we're nearly there. I would like to argue that:
1) All stuff in entrypoint is extra scripts 
2) All stuff in CMD (which _might_ point to an extra start script) is to start up the main app. 

End the entrypoint with exec "$@"

So you can have:
* If I just start the container, it will execute the entrypoint, and then the CMD
* If I do "docker run our_image_name $A_COMMAND"  it will execute the $A_COMMAND (usefull when debugging the image, and you do not want to start the specified commands)

It also nicely seperates the scripts from the app.

Let's also do what the rest of the world seems to do:
entrypoint script in /entrypoint.sh

If we need a startup script (eg apache + shib), put in in a start.sh script in /usr/local/sbin/start.sh 



This makes sure that

quartje commented:
This should work. I see that the entrypoint.sh scripts are not chmodded with an execute bit. Have you tested them that they work without explicit chmod?

danakim commented:
@quartje - it should work! Tested it. But just to make sure, I added an explicit chmod +x in all Dockerfiles, check my latest commit.

quartje commented on apache2-shibboleth/bin/entrypoint.sh:
```diff
@@ -35,12 +35,12 @@ for dir in \
     if [[ -v APACHE_UID_TO_CREATE ]]; then
         if [[ -v APACHE_GUID_TO_CREATE ]]; then
-            chown "$APACHE_UID_TO_CREATE:$APACHE_GUID_TO_CREATE" "$dir";
+            chown -R "$APACHE_UID_TO_CREATE:$APACHE_GUID_TO_CREATE" "$dir";
```
Only the cache directory should be recurively chown'd. The rest should not. The webserver should have very little write privileges on the local directories.

danakim replied:
I have left the "-R" because the only directories it actually runs these commands on is the cache directory, log directory and runtime. All of these should be recursively owned by Apache2.

But I have removed the actual application directory from this list, so it will have very few permissions there. It is all in my last commit.

quartje reviewed (changes requested):
I only want the cache dir to be writeable, not the other directories. I do not want the webserver to be able to write to other dirs than cache.

danakim commented:
@quartje fixed your latest request in the latest commit. Apache won't have write rights to the files / application directory. Only to the ones it needs: cache, logs and runtime (pid).

quartje reviewed (changes requested):
I have tested it, and there are still some loose ends:

1) The cache directory chowning is only on the php images. Adding those commands to php72-apache2 and php82-apache2 is enough.
2) The script should check if the cache dir is there. If not it should not chown (now I get an error about a missing cache dir)
3) The cache:clear command does not always exit correctly. The entrypoint.sh should use 
```
set -e
```
to make sure the script continues after errors
Also "exec" should not be used in front of the php bin/console cache:clear command

danakim commented:
@quartje - to answer your latest requests:

1. Right! No need to do that on the non-php images. I copied the script everywhere to keep it consistent, but it is not needed. Removed the cache command on the plain apache2 images.
2. I added a check for the directory.
3. I remember that that is the default bash behaviour: to continue on errors. But I think because we use it as an entrypoint on docker, it changes that behaviour as docker wants a clean run / exit. I have added the `set` command. But one note: `set -e` actually tells bash to exit immediately on any error. Basically the opposite of what you want. The command we want is `set +e` and I have added in the last commit.
